File: library/javascript/rtmsManager/FrontendWssManager.js
import { WebSocketServer } from 'ws';
import { FileLogger } from './utils/FileLogger.js';

export class FrontendWssManager {
  constructor(options = {}) {
    this.config = options.config || {};
    this.server = options.server || null;
    this.logger = options.logger || FileLogger;
    this.authorizeRegistration =
      options.authorizeRegistration || this.config.authorizeRegistration || null;
    this.frontendClients = new Set();
    this.broadcast = this.broadcastToFrontendClients.bind(this);
  }

  setup() {
    if (!this.server || this.config.frontendWssEnabled === false) {
      this.logger.log('[FrontendWssManager] 🧩 Frontend WSS skipped');
      return;
    }

    this.wss = new WebSocketServer({
      server: this.server,
      path: this.config.frontendWssPath,
      maxPayload: Number(this.config.frontendWssMaxPayloadBytes) || 64 * 1024
    });

    const pingInterval = 10000;
    this.pingTimer = setInterval(() => {
      const pingMsg = JSON.stringify({ type: 'ping' });
      for (const client of this.frontendClients) {
        if (client.readyState === 1) {
          client.send(pingMsg);
          // this.logger.log('[FrontendWssManager] 🔄 Ping sent to frontend client');
        }
      }
    }, pingInterval);

    this.wss.on('connection', (ws, request) => {
      this.frontendClients.add(ws);
      this.logger.log('[FrontendWssManager] 🌐 Frontend client connected (unregistered, registering now...)');

      // Kick if not registered within 15 seconds (increased from 5s to allow Zoom SDK context retrieval)
      const registrationTimeout = setTimeout(() => {
        if (!ws.meetingUUID || !ws.userID) {
          this.logger.log('[FrontendWssManager] ❌ Registration timeout. Closing connection.');
          if (ws.readyState === 1) {
            ws.send(JSON.stringify({ type: 'error', message: 'Registration timeout' }));
          }
          ws.terminate();
        }
      }, 15000);

      ws.send(JSON.stringify({ type: 'connected', message: 'Connected to RTMS backend. Please register.' }));

      ws.on('message', async (message) => {
        try {
          const data = JSON.parse(message.toString());
          
          if (data.type === 'pong') {
            // this.logger.log('[FrontendWssManager] 🏓 Pong received from frontend client');
            return;
          }

          if (data.type === 'register') {
            const { meetingUUID, userID, token } = data;
            const registration = await this.validateRegistration({
              meetingUUID,
              userID,
              token,
              request,
              socket: ws
            });

            if (registration.authorized) {
              ws.meetingUUID = String(registration.meetingUUID ?? meetingUUID);
              ws.userID = String(registration.userID ?? userID);
              clearTimeout(registrationTimeout);
              ws.send(JSON.stringify({
                type: 'registration_success',
                meetingUUID: ws.meetingUUID,
                userID: ws.userID
              }));
              this.logger.log(`[FrontendWssManager] ✅ Client registered: ${ws.userID} for meeting ${ws.meetingUUID}`);
            } else {
              this.logger.warn('[FrontendWssManager] Registration rejected');
              ws.send(JSON.stringify({ type: 'error', message: 'Registration unauthorized' }));
              ws.terminate();
            }
            return;
          }
        } catch (e) {
          this.logger.warn(`[FrontendWssManager] Invalid client message: ${e.message}`);
          if (ws.readyState === 1) {
            ws.send(JSON.stringify({ type: 'error', message: 'Invalid message' }));
          }
        }
      });

      ws.on('close', () => {
        clearTimeout(registrationTimeout);
        this.frontendClients.delete(ws);
        const info = ws.userID && ws.meetingUUID ? `: ${ws.userID} from ${ws.meetingUUID}` : '';
        this.logger.log(`[FrontendWssManager] ❌ Frontend client disconnected${info}`);
      });

      ws.on('error', (err) => {
        clearTimeout(registrationTimeout);
        this.frontendClients.delete(ws);
        this.logger.error('⚠️ Frontend WS error:', err);
      });
    });

    this.logger.log(`[FrontendWssManager] 🧩 Frontend WSS initialized at ${this.config.frontendWssPath}`);
  }

  async validateRegistration(context) {
    if (!context.meetingUUID || context.userID == null || !this.authorizeRegistration) {
      return { authorized: false };
    }

    const result = await this.authorizeRegistration(context);
    if (result === true) {
      return {
        authorized: true,
        meetingUUID: context.meetingUUID,
        userID: context.userID
      };
    }
    if (result && typeof result === 'object' && result.authorized === true) {
      return result;
    }
    return { authorized: false };
  }

  broadcastToFrontendClients(message) {
    const json = typeof message === 'string' ? message : JSON.stringify(message);
    for (const client of this.frontendClients) {
      if (client.readyState === 1 && client.meetingUUID && client.userID) {
        client.send(json);
      }
    }
  }

  /**
   * Broadcast to all clients in a specific meeting/session
   * @param {string} meetingUUID 
   * @param {Object|string} message 
   */
  broadcastToMeeting(meetingUUID, message) {
    const json = typeof message === 'string' ? message : JSON.stringify(message);
    for (const client of this.frontendClients) {
      if (client.readyState === 1 && client.meetingUUID === String(meetingUUID)) {
        client.send(json);
      }
    }
  }

  /**
   * Broadcast to a specific user in a specific meeting/session
   * @param {string} meetingUUID 
   * @param {string} userID 
   * @param {Object|string} message 
   */
  broadcastToUser(meetingUUID, userID, message) {
    const json = typeof message === 'string' ? message : JSON.stringify(message);
    for (const client of this.frontendClients) {
      if (
        client.readyState === 1 &&
        client.meetingUUID === String(meetingUUID) &&
        client.userID === String(userID)
      ) {
        client.send(json);
      }
    }
  }

  stop() {
    // Clear ping timer
    if (this.pingTimer) {
      clearInterval(this.pingTimer);
      this.pingTimer = null;
    }

    // Close all active client connections
    for (const client of this.frontendClients) {
      try {
        if (client.readyState === 1 || client.readyState === 0) { // OPEN or CONNECTING
          client.close(1000, 'Server shutting down');
        }
      } catch (err) {
        this.logger.error('[FrontendWssManager] Error closing client connection:', err);
      }
    }

    // Clear the clients set
    this.frontendClients.clear();

    // Close the WebSocket server
    if (this.wss) {
      try {
        this.wss.close(() => {
          this.logger.log('[FrontendWssManager] WebSocket server closed');
        });
        this.wss = null;
      } catch (err) {
        this.logger.error('[FrontendWssManager] Error closing WebSocket server:', err);
      }
    }

    this.logger.log('[FrontendWssManager] Stopped and cleaned up all connections');
  }
}

File: library/python/frontend_manager/frontend_wss_manager.py
import asyncio
import json
from typing import Set, Dict, Any, Optional
from dataclasses import dataclass, field

from ..rtms_manager.utils.logger import FileLogger


@dataclass
class FrontendClient:
    websocket: Any
    meeting_uuid: Optional[str] = None
    user_id: Optional[str] = None
    registered: bool = False


class FrontendWssManager:
    def __init__(self, wss_path: str = '/ws', ping_interval: int = 10):
        self.wss_path = wss_path
        self.ping_interval = ping_interval
        self.clients: Set[FrontendClient] = set()
        self._ping_task: Optional[asyncio.Task] = None
        self._server = None

    async def setup_with_websockets(self, server):
        self._server = server
        self._ping_task = asyncio.create_task(self._ping_loop())
        FileLogger.log(f'[FrontendWssManager] WebSocket server initialized at {self.wss_path}')

    async def handle_connection(self, websocket, path: str = ''):
        if path != self.wss_path and self.wss_path not in path:
            return

        client = FrontendClient(websocket=websocket)
        self.clients.add(client)
        FileLogger.log('[FrontendWssManager] Frontend client connected (unregistered)')

        registration_timeout = 15
        
        async def check_registration():
            await asyncio.sleep(registration_timeout)
            if not client.registered:
                FileLogger.log('[FrontendWssManager] Registration timeout. Closing connection.')
                await self._send_to_client(client, {'type': 'error', 'message': 'Registration timeout'})
                await websocket.close()

        asyncio.create_task(check_registration())

        await self._send_to_client(client, {
            'type': 'connected', 
            'message': 'Connected to RTMS backend. Please register.'
        })

        try:
            async for message in websocket:
                try:
                    data = json.loads(message)
                except json.JSONDecodeError:
                    continue

                if data.get('type') == 'pong':
                    continue

                if data.get('type') == 'register':
                    meeting_uuid = data.get('meetingUUID')
                    user_id = data.get('userID')

                    if meeting_uuid and user_id:
                        client.meeting_uuid = meeting_uuid
                        client.user_id = user_id
                        client.registered = True
                        await self._send_to_client(client, {
                            'type': 'registration_success',
                            'meetingUUID': meeting_uuid,
                            'userID': user_id
                        })
                        FileLogger.log(f'[FrontendWssManager] Client registered: {user_id} for meeting {meeting_uuid}')
                    else:
                        FileLogger.log('[FrontendWssManager] Registration rejected: Invalid meetingUUID or userID')
                        await self._send_to_client(client, {'type': 'error', 'message': 'Registration invalid'})
                        await websocket.close()

        except Exception as e:
            FileLogger.error(f'[FrontendWssManager] WebSocket error: {e}')
        finally:
            self.clients.discard(client)
            info = f': {client.user_id} from {client.meeting_uuid}' if client.registered else ''
            FileLogger.log(f'[FrontendWssManager] Frontend client disconnected{info}')

    async def _send_to_client(self, client: FrontendClient, message: Dict[str, Any]):
        try:
            await client.websocket.send(json.dumps(message))
        except Exception:
            pass

    async def _ping_loop(self):
        while True:
            await asyncio.sleep(self.ping_interval)
            ping_msg = json.dumps({'type': 'ping'})
            for client in list(self.clients):
                try:
                    await client.websocket.send(ping_msg)
                except Exception:
                    self.clients.discard(client)

    def broadcast(self, message: Dict[str, Any]):
        asyncio.create_task(self._broadcast_async(message))

    async def _broadcast_async(self, message: Dict[str, Any]):
        msg_json = json.dumps(message)
        for client in list(self.clients):
            try:
                await client.websocket.send(msg_json)
            except Exception:
                self.clients.discard(client)

    def broadcast_to_meeting(self, meeting_uuid: str, message: Dict[str, Any]):
        asyncio.create_task(self._broadcast_to_meeting_async(meeting_uuid, message))

    async def _broadcast_to_meeting_async(self, meeting_uuid: str, message: Dict[str, Any]):
        msg_json = json.dumps(message)
        for client in list(self.clients):
            if client.meeting_uuid == meeting_uuid:
                try:
                    await client.websocket.send(msg_json)
                except Exception:
                    self.clients.discard(client)

    def broadcast_to_user(self, meeting_uuid: str, user_id: str, message: Dict[str, Any]):
        asyncio.create_task(self._broadcast_to_user_async(meeting_uuid, user_id, message))

    async def _broadcast_to_user_async(self, meeting_uuid: str, user_id: str, message: Dict[str, Any]):
        msg_json = json.dumps(message)
        for client in list(self.clients):
            if client.meeting_uuid == meeting_uuid and client.user_id == user_id:
                try:
                    await client.websocket.send(msg_json)
                except Exception:
                    self.clients.discard(client)

    async def stop(self):
        if self._ping_task:
            self._ping_task.cancel()
            try:
                await self._ping_task
            except asyncio.CancelledError:
                pass

        for client in list(self.clients):
            try:
                await client.websocket.close()
            except Exception:
                pass

        self.clients.clear()
        FileLogger.log('[FrontendWssManager] Stopped and cleaned up all connections')
